File: 常考代码/笔试题/京东笔试.js
let line = '1000 100'
line = line.split(" ").map(item=>Number(item))

console.log(line)

function minimalPower(arr) {
  n = arr.length;
  arr.sort((a,b)=>{
      if(a[0] == b[0]){
          return b[1] - a[1]
      }else{
          return a[0] - b[0]
      }
  })
  let res = 0;
  for(let i = n-1;i>=0;i--){
      if(i === n-1){
          res = arr[i][0];
          continue;
      }else{
          res = Math.max(res - arr[i][1], arr[i][0]);
      }
  }
  return res;
}


const arr = [[1000, 100], [100, 100], [200, 700],[100,200]];
console.log(minimalPower(arr))

File: 常考代码/笔试题/京东笔试.py
def csort(a,b):
    if a[0] == b[0]:
        return 
def minimalPower(arr):
    arr.sort(key = lambda a: a[0])
    gainPower = 0
    maxPower = max([item[0] for item in arr])
    minPower = min([item[0] for item in arr])
    needPower = 0
    for i in range(len(arr)-1): # 能获得的战斗力应该排除战斗力最强boss所能获得的战斗力
        gainPower +=arr[i][1]
    # 如果能够获得的战斗力已经大于最大的战斗力，则初始战斗力只需要为最小的战斗力
    if gainPower >= maxPower:
        needPower = minPower
    # 如果获得的战斗力不大于最大的战斗力，则需要补齐,且需要和最小战斗力进行比较，选取更大的那个
    elif gainPower < maxPower:
        needPower = maxPower - gainPower if maxPower - gainPower >= minPower else minPower
    return needPower

arr = [[1000,100], [100,100],[200,700]]

print(minimalPower(arr))